src/mpsc/async_impl.rs

```diff
@@ -281,6 +281,161 @@ feature! {
         }
     }
 
+    feature! {
+        #![feature = "stream"]
+        use futures::sink::Sink;
+        use std::mem;
+
+        /// PollSender internal state.
+        enum PollSenderState<'a, T, R> {
+            Idle(Sender<T, R>),
+            Acquiring(Pin<Box<SendRefFuture<'a ,T, R>>>),
+            ReadyToSend(SendRef<'a, T>),
```

Comment:
it feels a little strange to me that the `poll_reserve` method stores the `SendRef` inside the `PollSender` and then consumes it when `send` is called, rather than just having `poll_reserve` return a `SendRef` to the user that can then be used to write a value. this API would allow the user to write to the slot in place through the `SendRef`, which isn't currently possible with the current API.

OTTOH, we would still need to store the `SendRef` internally in order to make the `Sink` impl work, but I think we could change the `poll_reserve` function to return a `SendRef`, and only store that `SendRef` internally when `Sink::poll_ready` is called?  Of course, in order for that to work, we would need to add an owned `SendRef` variant, as you mentioned. 

This would make the API somewhat different from `tokio_util::sync`'s `PollSender`, but I think it's worth it to allow the user to actually write to a slot in place through a `SendRef`. If we wanted to present an identical API to `tokio_util::sync::PollSender`, we could also add a new poll method that returns a `SendRef`, and have `poll_reserve` just call that method and then store the returned `SendRef` if it returns `Poll::Ready`...what do you think?

Reply:
Ok. I used the `PollSender` approach in order to not invent something new for something in place - but yes - the ability to directly return the `SendRef` is plausible.  Having a different implementation than `PollSender` is not a problem for me personally because I assume that in the "normal" cases the `Stream`/`Sink` API is used.
 If anyone calls `poll_reserve`, they normally know that they do and a difference to `PollSender` doesn't matter at all.